sql/core/src/test/scala/org/apache/spark/sql/execution/ReduceNumShufflePartitionsSuite.scala

```diff
@@ -587,4 +587,22 @@ class ReduceNumShufflePartitionsSuite extends SparkFunSuite with BeforeAndAfterA
     }
     withSparkSession(test, 200, None)
   }
+
+  test("Union two datasets with different pre-shuffle partition number") {
+    val test: SparkSession => Unit = { spark: SparkSession =>
+      val df1 = spark.range(3).join(spark.range(3), "id").toDF()
+      val df2 = spark.range(3).groupBy().sum()
+
+      val resultDf = df1.union(df2)
+
+      checkAnswer(resultDf, Seq((0), (1), (2), (3)).map(i => Row(i)))
```

Comment:
does this fail without the fix?

Reply:
It does. The plan is:
```
AdaptiveSparkPlan(isFinalPlan=false)
+- Union
   :- Project [id#0L]
   :  +- SortMergeJoin [id#0L], [id#2L], Inner
   :     :- Sort [id#0L ASC NULLS FIRST], false, 0
   :     :  +- Exchange hashpartitioning(id#0L, 5), true
   :     :     +- Range (0, 3, step=1, splits=12)
   :     +- Sort [id#2L ASC NULLS FIRST], false, 0
   :        +- Exchange hashpartitioning(id#2L, 5), true
   :           +- Range (0, 3, step=1, splits=12)
   +- HashAggregate(keys=[], functions=[sum(id#6L)], output=[sum(id)#10L])
      +- Exchange SinglePartition, true
         +- HashAggregate(keys=[], functions=[partial_sum(id#6L)], output=[sum#14L])
            +- Range (0, 3, step=1, splits=12)
```
and the error comes from this assert: https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ReduceNumShufflePartitions.scala#L136
Honestly I'm not sure why the `Exchange SinglePartition` has `true`

Reply:
Can you fill the `Does this PR introduce any user-facing change` section? Changing a query from failure to runnable is a user-facing change.

Reply:
Oh ok, sure, filled.